onnx_trt_backend.cpp

```diff
@@ -0,0 +1,749 @@
+    return ONNXIFI_STATUS_INVALID_DATATYPE;
+  }
+  // Check input shape
+  auto ret = CheckShape(dims, tensor);
+  if (ret != ONNXIFI_STATUS_SUCCESS) {
+    return ret;
+  }
+
+  // For CPU tensor, we need to create a device memory and the bind. For CUDA
+  // tensor, we can bind directly
+  if (tensor.memoryType == ONNXIFI_MEMORY_TYPE_CPU) {
+    void *cuda_buffer;
+    size_t footprint = GetTensorFootprint(tensor);
+    if (!footprint) {
+      return ONNXIFI_STATUS_INVALID_SHAPE;
+    }
+    if (cudaMalloc(&cuda_buffer, footprint) != cudaSuccess) {
+      return ONNXIFI_STATUS_NO_DEVICE_MEMORY;
+    }
+    device_buffers_.emplace(tensor.name, cuda_buffer);
+    bindings_.push_back(cuda_buffer);
+  } else {
+    bindings_.push_back((void *)(tensor.buffer));
+  }
+
+  return ONNXIFI_STATUS_SUCCESS;
+}
+
+onnxStatus GraphRep::InitIO(uint32_t inputsCount,
+                            const onnxTensorDescriptor *inputDescriptors,
+                            uint32_t outputsCount,
+                            const onnxTensorDescriptor *outputDescriptors) {
+  CudaDeviceGuard guard(device_id_);
+  ClearDeviceBuffers();
+  // Setup the input/output bindings and decide batch size
+  for (unsigned i = 0; i < inputsCount; ++i) {
+    if (!inputDescriptors[i].name) {
+      return ONNXIFI_STATUS_INVALID_NAME;
+    }
+    // We only support NCHW
+    if (inputDescriptors[i].dimensions != 4) {
+      return ONNXIFI_STATUS_INVALID_SHAPE;
+    }
+    if (i == 0) {
+      batch_size_ = inputDescriptors[i].shape[0];
+    } else {
+      if (batch_size_ != inputDescriptors[i].shape[0]) {
+        return ONNXIFI_STATUS_INVALID_SHAPE;
+      }
+    }
+    std::cerr << "Adding input " << i << ": " << inputDescriptors[i].name
+              << ", type: " << inputDescriptors[i].memoryType << std::endl;
+    input_map_.emplace(std::string(inputDescriptors[i].name),
+                       inputDescriptors + i);
+  }
+
+  // We don't support the case when batch size is larger than max batch size
+  // yet, but this is not a hard constraint.
+  if (batch_size_ > max_batch_size_) {
+    return ONNXIFI_STATUS_NO_DEVICE_RESOURCES;
+  }
+
+  for (unsigned i = 0; i < outputsCount; ++i) {
+    if (!outputDescriptors[i].name) {
+      return ONNXIFI_STATUS_INVALID_NAME;
+    }
+    // We only support NCHW
+    if (outputDescriptors[i].dimensions != 4) {
+      return ONNXIFI_STATUS_INVALID_SHAPE;
+    }
+    if (batch_size_ != outputDescriptors[i].shape[0]) {
+      return ONNXIFI_STATUS_INVALID_SHAPE;
+    }
+    output_map_.emplace(std::string(outputDescriptors[i].name),
+                        outputDescriptors + i);
+  }
+
+  int nbindings = trt_engine_->getNbBindings();
+  for (int b = 0; b < nbindings; ++b) {
+    nvinfer1::Dims dims = trt_engine_->getBindingDimensions(b);
+    // Check data type consistency
+    auto binding_datatype = trt_engine_->getBindingDataType(b);
+    if (binding_datatype != nvinfer1::DataType::kFLOAT) {
```

Comment:
Does TensorRT support any operations with non-float types?

Reply:
TensorRT supports fp16 and int8. But it's not readily supported in onnx-tensorrt yet AFAIK.